libraries/AP_HAL_ChibiOS/UARTDriver.cpp

```diff
@@ -529,12 +569,46 @@ void UARTDriver::dma_tx_deallocate(Shared_DMA *ctx)
     txdma = nullptr;
     chSysUnlock();
 }
+#endif
+
+// transfer a frame from the read buffer to the frame buffer
+void UARTDriver::transfer_frame()
+{
+    uint32_t fsize = _readbuf.available();
+
+    // no data
+    if (fsize == 0) {
+        return;
+    }
+
+    // framing error, restart at the next frame
+    if (fsize > frame_size) {
+        _frame_readbuf.clear();
+        _readbuf.clear();
+        return;
+    }
+
+    // read the remaining available bytes as a frame
+    _frame_readbuf.write((const uint8_t*)&fsize, sizeof(uint32_t));
+    while (fsize > 0) {
```

Comment:
nasty problem if another thread does a read, we could end up with fsize wrapping, and we fall out of the sky with the while loop

Reply:
I've now locked the kernel around this call - I don't think another thread can read until we are done

Reply:
I've now locked the kernel around this call so I don't think another thread can read while the transfer is in progress

Reply:
Removed